src/components/fields/ObjectField.js

```diff
@@ -168,14 +168,33 @@ class ObjectField extends Component {
   }
 
   handleAddClick = schema => () => {
-    const type = schema.additionalProperties.type;
+    let type = schema.additionalProperties.type;
     const newFormData = { ...this.props.formData };
+
+    if (
+      type === undefined &&
+      schema.additionalProperties.hasOwnProperty("$ref")
+    ) {
+      const refSchema = retrieveSchema(
+        schema.additionalProperties,
+        this.getDefinitions(),
+        this.props.formData
+      );
+      type = refSchema.type;
+    }
+
     newFormData[
       this.getAvailableKey("newKey", newFormData)
     ] = this.getDefaultValue(type);
+
     this.props.onChange(newFormData);
   };
 
+  getDefinitions() {
+    const { registry = getDefaultRegistry() } = this.props;
```

Comment:
Since getDefinitions() is used only once, can you inline it? (unless you think there might be a better reason to keep it as its own function)